Add utility functions to rollcage

ProjectFrank opened the pull request:
This PR adds three functions that consumers of the rollcage library would likely need to implement themselves: 

- ring middleware for handling uncaught exceptions thrown from ring handlers
- a function that returns a configured reporter that reports exceptions to rollbar
- a configuration function that creates a rollbar client and sets up rollcage as the default uncaught exception handler

ProjectFrank committed:
Add ring middleware for handling uncaught exceptions

ProjectFrank committed:
Add function for configuring rollcage

ProjectFrank committed:
Add rollbar-reporter helper function

ProjectFrank committed:
Upgrade bond dependency

marcomorain commented on src/circleci/rollcage/ring_middleware.clj:
```diff
@@ -0,0 +1,12 @@
+        (handler req)
+        (catch Exception e
+          (rollcage/error rollcage-client e (select-keys req [:uri]))
+          (throw e))))))
```
The `if` makes control flow harder to follow imo - what about: 

```Clojure
(defn wrap-rollbar [handler rollcage-client]
  (fn [req]
    (try
      (handler req)
      (catch Exception e
        (when rollcage-client
          (rollcage/error rollcage-client e (select-keys req [:uri]))
        (throw e))))))
```

ProjectFrank replied:
I would have to disagree. I don't think this kind of control flow is uncommon when it comes to ring middleware, or higher order functions in general.

It's hardly a performance improvement, but isn't it still better to have the conditional logic evaluate once when the handler is created, as opposed to each time the returned handler is invoked?

marcomorain replied:
> It's hardly a performance improvement, but isn't it still better to have the conditional logic evaluate once when the handler is created, as opposed to each time the returned handler is invoked?

That's a good point!

marcomorain commented on src/circleci/rollcage/ring_middleware.clj:
```diff
@@ -0,0 +1,12 @@
+(ns circleci.rollcage.ring-middleware
+  (:require [circleci.rollcage.core :as rollcage]))
+
+(defn wrap-rollbar [handler rollcage-client]
```
This is nice to have in Rollcage :+1:

marcomorain commented on src/circleci/rollcage/core.clj:
```diff
@@ -187,3 +188,24 @@
+
+(defn configure-rollbar!
+  "Make rollcage handle uncaught exceptions and return a rollcage client."
+  [rollbar-token environment]
```
I don't think this function adds value to Rollcage - I'd rather see these lines of code explicitly in the `main` function of each project. I could be convinced otherwise.

One thing I don't like is that the call can only specify one parameter `environment` - if the user wanted to specify `:code-version` too,  then we would have to change this function. It would be better to take all client options as the second param.

But I think this function belongs explicitly in each app tbh.

marcomorain commented on src/circleci/rollcage/core.clj:
```diff
@@ -187,3 +188,24 @@
+  [rollcage-client params]
+  (infof "rollbar reporting configured for %s" params)
+  (fn [^Throwable ex]
+    (try
```
I think the try/catch here belongs in Rollcage itself - otherwise every client if going to have to implement it. We should change it so that it doesn't throw exceptions. How about we add try/catch to `send-item`? We will need to think about the return value of `send-item`.

marcomorain commented on src/circleci/rollcage/core.clj:
```diff
@@ -187,3 +188,24 @@
 (def info     (partial notify "info"))
 (def debug    (partial notify "debug"))
+
+(defn rollbar-reporter
```
If we move the try/catch code into `send-item` then we don't need `rollbar-reporter` - since it can then be replaced with `(partial rollcage/notify "error" client)`.

marcomorain commented on src/circleci/rollcage/core.clj:
```diff
@@ -1,6 +1,7 @@
 (ns circleci.rollcage.core
   (:require
     [clojure.string :as string]
+    [clojure.tools.logging :refer (infof errorf)]
```
Let's not bring this extra dependency into Rollcage.

I would rather do this by supplying an optional `failure-fn` to the client that will be called with failures.

```Clojure
(defn send-item [endpoint item]
  (try
    (let [result (post endpoint {:body (json/generate-string item {:key-fn snake-case})
                                 :content-type :json})]
      (json/parse-string (:body result) true))
    (catch Exception e
      (when-let [failure-fn (:failure-fn client)]
        (failure-fn item e))

ProjectFrank committed:
Add :failure-fn key to rollcage client map

This adds an option to create a rollcage client with an error handler
for handling exceptions that arise when sending messages to rollbar.

ProjectFrank committed:
Remove rollbar-reporter-fn

ProjectFrank committed:
Remove configure-rollbar! fn

ProjectFrank committed:
Fix require in middleware ns

ProjectFrank committed:
Update ring middleware test

marcomorain committed:
Replace failure-fn with a generic result-fn

Call result-fn after anything is sent to Rollbar.

marcomorain commented:
@ProjectFrank I pushed a commit on top that replaces failure-fn with a generic result-fn that is called in all cases - success or failure. Client code can use the `:err` key to check for errors - https://rollbar.com/docs/api/items_post/#success

what do you think?

ProjectFrank commented:
I agree that having a generic `result-fn` is definitely more powerful than just a `failure-fn`. I think that the `result-fn` should still rethrow the exception, especially since the default `result-fn` of `(constantly nil)` eats them without any kind of handling. Thoughts?

ianconsolata commented:
@ProjectFrank per our discussion this morning, can you revive and finish this PR? Building another service that will want this functionality.

ProjectFrank commented:
@marcomorain bump

marcomorain commented:
> I agree that having a generic result-fn is definitely more powerful than just a failure-fn. I think that the result-fn should still rethrow the exception, especially since the default result-fn of (constantly nil) eats them without any kind of handling. Thoughts?

I don't think we should throw if sending an exception to Rollbar fails. I don't know what the application could do in response to an exception being throw?

marcomorain commented:
@ProjectFrank - sorry I forgot about this PR - I'm working on it now.
